Jk/cumulus 3320 upgrade docs (#3696)
nasa/cumulus: docs/upgrade-notes/upgrade-execution-table-CUMULUS-3320.md

---
id: upgrade_execution_table_CUMULUS_3320
title: Upgrade execution table to CUMULUS-3320
hide_title: false
---

# Background

As part of the performance/feature evaluation work in completing the update requested as part of CUMULUS-3320, several updates were required to make `executions` table deletes/reads more performant.

These changes required creating new indexes and modifying a table constraint - operations which can take some time and require manual upgrade steps for database deployments where the updates will exceed the bootstrap lambda's 15 minute timeout.

 Please note that testing in a production-similar environment is *strongly* advised if you are concerned about precise query times and/or downtime.

The following procedures detail how this upgrade may be performed:

## Adding the `executions_parent_cumulus_id_index` index

### Reasoning

> This index is required to support effective deletion of records given query analysis implies the same table foreign-key constraint results in table scan to enforce the constraint for each deletion.

Reviewer: ```suggestion
This index is required to support effective deletion of records. Our query analysis implies the same table foreign-key constraint results in a slow table scan to enforce the constraint for each deletion.
```

#### Utilize the normal cumulus deployment

The cumulus module will run the migrations on the database when it's deployed.   For databases with smaller holdings and not under heavy load such that the sum of the migrations to be run will complete in < 15 minutes, the migration will run the following SQL query:

```sql
CREATE INDEX CONCURRENTLY IF NOT EXISTS executions_parent_cumulus_id_index ON executions(parent_cumulus_id)'
```

#### Manually create the index

##### Create index with ingest halted:

The recommended approach to create the index is to halt ingest activity that requires write access to the execution table, then run the query to create the index.  This will result in a much faster index operation than attempting to create the index concurrently, however the table will be locked until the index is complete.

For reference, in testing, using a 4ACU Aurora Serverless V1 cluster on a table with 15 million rows this migration took roughly 2 minutes to complete.

To do this, use the following query:

```sql
CREATE INDEX executions_parent_cumulus_id_index ON executions(parent_cumulus_id)'
```

##### Create index concurrently/with ongoing:

> The required index can also be created while the database is in use prior to installing the upgrade containing CUMULUS-3320 by running the following query:

Reviewer: ```suggestion
The required index can also be created while the database is in use *prior to installing the upgrade containing CUMULUS-3320* by running the following query:
```

#### Utilize the normal cumulus deployment

The cumulus module will run the migrations on the database when it's deployed.   For databases with smaller holdings and not under heavy load such that the sum of the migrations to be run will complete in < 15 minutes, the migration will run the following SQL query:

```sql
CREATE INDEX CONCURRENTLY IF NOT EXISTS executions_collection_cumulus_id_index ON executions(collection_cumulus_id)
```

#### Manually create the index

##### Create index with ingest halted:

The recommended approach to create the index is to halt ingest activity that requires write access to the execution table, then run the query to create the index.  This will result in a much faster index operation than attempting to create the index concurrently, however the table will be locked until the index is complete.

For reference, in testing, using a 4ACU Aurora Serverless V1 cluster on a table with 15 million rows this migration took roughly 3 minutes to complete.

To do this, use the following query:

```sql
CREATE INDEX executions_collection_cumulus_id_index ON executions(collection_cumulus_id)
```

> The required index can also be created while the database is in use prior to installing the upgrade containing CUMULUS-3320 by running the following query:

Reviewer: ```suggestion
The required index can also be created while the database is in use *prior to installing the upgrade containing CUMULUS-3320* by running the following query:
```

```sql
ALTER TABLE executions DROP CONSTRAINT IF EXISTS executions_parent_cumulus_id_foreign
ALTER TABLE executions ADD CONSTRAINT executions_parent_cumulus_id_foreign FOREIGN KEY (parent_cumulus_id) REFERENCES executions(cumulus_id) ON DELETE SET NULL';
```

#### Manually update the constraint

To manually update the constraint, do the following, in order:

##### Modify the existing constraint without validation

The following query will, in a transaction, remove the existing constraint and re-enable the updated constraint.   Adding `NOT VALID` to the creation will cause postgres not to validate values in existing rows, but will enforce them for new rows.    This *greatly* reduces downtime requirements as the constraint can be validated while the table is in use, and this transaction should be extremely quick:

```sql
BEGIN;
ALTER TABLE executions DROP CONSTRAINT IF EXISTS executions_parent_cumulus_id_foreign;
ALTER TABLE executions ADD CONSTRAINT executions_parent_cumulus_id_foreign FOREIGN KEY (parent_cumulus_id) REFERENCES executions(cumulus_id) ON DELETE SET NULL NOT VALID;
END;
```

#### Re-validate the constraint

> This query can be run at any time following the above step and will not lock the full table for new row inserts/selects:

Reviewer: Maybe a note about why we need to validate? Or is that too close to 'how to SQL'?

Reviewer: Eh, it's mostly explained above.

Author: There's definitely an audience background concern here.   I think I fall on the side of YAGNI on this, but if you feel strongly I can try to include some links to postgres documents.
